### PR TITLE
Add thread pools for gc and compaction jobs

### PR DESCRIPTION
**Pull request by bojanserafimov**
Resolves https://github.com/neondatabase/neon/issues/1646

I still have a good amount of cleanup+simplification+documentation to do but feel free to comment on the approach and the logic inside `worker_main`.

### CODE REVIEW

**Review by hlinnaka — COMMENTED** [review]
Needs comments as you said. But other than that, looks good.

So with this, there is a separate thread pool and separate job queue for compaction and GC jobs? (I don't know if that's better or worse than a single pool, just observing)

**Comment by bojanserafimov** [comment]
> Needs comments as you said. But other than that, looks good.
> 
> So with this, there is a separate thread pool and separate job queue for compaction and GC jobs? (I don't know if that's better or worse than a single pool, just observing)

My implementation is generic over `Job` just to make tests easier (will add a few more tests). I'm not opinionated yet on whether the queues should be separate. It's easy to merge them by defining `enum TenantJob { GcJob(Tenant), CompactionJob(Tenant)}`. Though separation might decrease the blast radius of bugs.. not sure yet.

File: pageserver/src/tenant_jobs/worker.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                panic!("AAA");
+            }
+            println!("{}", self.to_print);
+            Ok(Some(Instant::now().add(Duration::from_millis(10))))
+        }
+    }
+
+    static TEST_POOL: OnceCell<Pool<PrintJob>> = OnceCell::new();
+
+    #[tokio::test]
```
Comment (LizardWizzard) [inline pageserver/src/tenant_jobs/worker.rs:126]: Is there any reason to use tokio testing infra? There is no async code involved besides time::sleep which has a non async counterpart in std: https://doc.rust-lang.org/stable/std/thread/fn.sleep.html#

**Review by LizardWizzard — COMMENTED** [review]
Some concerns:

* Do not see a way to remove a particular job from the scheduling, that would be important when we detach timelines because detached timelines should not leave traces in internal structures and pile up indefinitely.
* Any reason to avoid existing libraries? Just to consider pros/cons of existing solutions vs our own implementation. For example, [rayon](https://rust-lang.github.io/hashbrown/rayon/index.html) is a well-known library used to parallelize things, probably even a go-to threadpool impl.
* Currently there is no way to increase the number of workers. An existing library can give that for free.
* Was tokio considered as an option? Async adoption in our codebase grows with recent changes to walreceiver, and it would be a bit weird to have one more threadpool/scheduler in the codebase without a good reason.

* How do we deal with stuck jobs? E g if a job deadlocks inside is there a way to cancel it? Will it affect other tenants? Proper cancellation support can require quite a lot of effort, which reminds me of the cancellation story in async. So it looks like we're reimplementing parts of it without using the already built and battle-tested infrastructure.

Thoughts?

**Comment by bojanserafimov** [comment]
> 

Good questions. I'm open to suggestions, but here's my reasoning of why I went with this approach:

Rayon seems to be great for parallelizing a big CPU-heavy task on all your cores, to finish it faster. That's not our use case. I'm not too familiar with the library, pls correct me if there are features I'm overlooking.

Async is great for dealing with unpredictable concurrency and arbitrary number of tasks. That's also not our use case. Also:
1. An async transition is a whole big project
2. Limiting the entire pageserver to non-blocking code seems like a risk we don't have to take
3. Not sure if the tokio executer will give us the cross-tenant fairness we need
4. Observability/debuggability

For these kind of "tenant maintenance jobs", we want to:
1. Maximize observability
2. Make sure to approximately hit deadlines
3. Schedule fairly across tenants
4. Limit blast radius of bugs
5. Schedule to optimize for best SSD utilization
6. Schedule to prevent overallocation that could lead to perf degradation (or variability) elsewhere

A custom implementation seems like the simplest solution. It's not that much code, but as you said it has some problems. To answer your questions:
1. Currently jobs remove themselves when they notice the tenant is not active. Maybe eager removal is better. I left a todo and I'm still debating.
2. You can increase the number of workers by spawning more threads. Not sure we'd want to do that any time soon though. Will we ever need more than 50 threads? There's only so much SSD to go around, and 50 threads is already better than 1 thread per tenant (current state of things).
3. On stuck jobs: Making this situation detectable/alertable is on my list. If a task is taking longer than it's scheduling period, that should be suspicious. Now if you're sure there's some sort of deadlock, you can kill the worker and mark the task broken. Though it would be hard to gracefully kill the thread, since if it's stuck it's likely not checking for shutdown requests. I'm open to suggestions.

File: pageserver/src/tenant_jobs/worker.rs (hunk 1 above)
Comment (bojanserafimov) [reply pageserver/src/tenant_jobs/worker.rs:126]: no reason. It will go away.

**Comment by LizardWizzard** [comment]
Regarding rayon, it has ThreadPoolBuilder and an ability to spawn closures into it, which might help to avoid writing custom code to spawn workers and distribute tasks across them. 

> Async is great for dealing with unpredictable concurrency and arbitrary number of tasks. That's also not our use case.

But why? Both multithreaded async scheduler and threadpool are examples of M:N concurrency models, arent they? 

> An async transition is a whole big project

I agree, but we use async already in quite a lot of places. Plus tokio has spawn_blocking (which has a threadpool underneath) for places that we do not want to fully convert into async fashion.

> Limiting the entire pageserver to non-blocking code seems like a risk we don't have to take

Keeping both in parallel is a risk as well. And as I said above, it is not necessary to convert everything to async. Some parts can stay sync if they need to.

> Not sure if the tokio executer will give us the cross-tenant fairness we need

But will threadpool give us that? Tokio and async in general give quite powerful building blocks that we can use to implement the fairness guarantees we need. In general async gives more control over tasks, because they're scheduled in user space. There is an interesting example from glommio framework. It has multiple task queues and you can set shares between them (and even set latency targets, so the executor can distinguish latency-sensitive tasks and e g background activity), and choose which queue to use to spawn tasks into. See the scheduling section here https://docs.rs/glommio/latest/glommio/#scheduling 

Glommio represents quite a specific paradigm that requires writing code specifically for that paradigm (everything is thread-local as much as possible, no arcs, no cross-thread mutexes, message passing, etc). So it is not the shift we can make now. As it involves huge code changes.

> Observability/debuggability

IMO tracing gives reasonable observability. Async is indeed harder to debug from a debugger (e g deadlocks are pain that is somewhat softened with a tokio-console which is an additional tool that needs instrumentation etc etc.) so this is a trade-off.

> Though it would be hard to gracefully kill the thread, since if it's stuck it's likely not checking for shutdown requests. I'm open to suggestions.

It is almost impossible to kill a thread without its assistance. The answer to that IMO is async cancellation.

IMO with the list of desired features, it seems that we're building our own async executor but without async :) 

> Currently jobs remove themselves when they notice the tenant is not active.

I think this condition needs to be tweaked, but this is not a problem of this PR. GC needs to stop itself when there is nothing to collect and no walreceiver (tenant is Idle) and it should wakeup when PITR interval is reached so something may get old enough to be cleaned up. Or if everything is compacted all the data is not historical it can stop unconditionally until tenant becomes active again.


> You can increase the number of workers by spawning more threads.

Oh, my bad, initially I've missed it. But anyway would be interesting to see how it behaves with the amount of tenants that is  generated in the capacity test environment.

> Schedule to optimize for best SSD utilization (out of scope in this PR)

Curious about what is behind that :) Do you mean io_uring? Or something else? io_uring looks promising and in general, it favors the async paradigm. There is [tokio_uring](https://github.com/tokio-rs/tokio-uring) and glommio is fully based on io_uring capabilities. There is a good introductory post for glommio, may be an interesting reading (for me it was): https://www.datadoghq.com/blog/engineering/introducing-glommio/. 

You know that I like the async paradigm despite its caveats for greater flexibility and greater control. For sure it has problems like any other solution (there is no silver bullet as always). I'm not insisting on rewriting everything in async and do not say that it will magically solve all our problems. But it provides building blocks that can help us to achieve our goals. 

IMO we need a strong concurrency framework as a foundation for our service to be robust and reliable. Thread pool only for GC and compaction solves the local problem. A significant portion of threads come from our networking implementation. Do we need a separate thread pool here or another tokio runtime? This also needs time to investigate performance consequences. Not sure that we have the capacity for it now (or have we?). Another thing that we have is a remote storage sync thread that executes s3 synchronization tasks and already implements a retry policy. Once we dig more into implementing on-demand layer upload/download we will need to bridge the gap between sync and async again. 

Do we fairly slice pageserver resources for tenants? If I'm understanding it correctly the one who will have more `pagestream` connections will take more pageserver CPU time and will have more cache/memory dedicated to it. Is it ok or we should do something about it?

Another article I wanted to share: https://www.influxdata.com/blog/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/  TLDR how to use a pair of tokio runtimes for different kinds of tasks.

Basically, my question is what is the direction in which we're going?

**Comment by bojanserafimov** [comment]
@LizardWizzard Thanks for the links, reading.. :)

Tokio spawn_blocking seems promising. It could simplify things. Though there's no async cancellation for spawn_blocking.

**Comment by LizardWizzard** [comment]
> Though there's no async cancellation for spawn_blocking

Yeah, because it runs blocking code on a threadpoll threads and thus has the same problems, but it combines more or less nicely with other async stuff

**Comment by hlinnaka** [comment]
I think the approach in this PR is good. async vs sync is a different discussion, and is out of scope of this PR. The current model in the pageserver is that most code is synchronous, with the exception of a few limited places that need to wait for many things at the same time, which use async with the current-thread executor. This PR is in line with that.

**Comment by bojanserafimov** [comment]
@hlinnaka here's a draft for the tokio approach https://github.com/neondatabase/neon/compare/tenant-tasks?expand=1

It's the same solution as this PR (gc jobs are still blocking, and running on dedicated threads), but with less room for bugs.

**Comment by bojanserafimov** [comment]
going with this instead -> https://github.com/neondatabase/neon/pull/1933
